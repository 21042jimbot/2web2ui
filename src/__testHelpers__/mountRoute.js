import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import axios from 'axios';
import { Provider } from 'react-redux';

import configureStore from 'src/store';

import App from 'src/App';
import { authenticate } from 'src/actions/auth';
import { initializeAccessControl } from 'src/actions/accessControl';

import asyncFlush from 'src/__testHelpers__/asyncFlush';

const forceUpdate = async (wrapper) => {
  await asyncFlush();
  wrapper.update();
};

const simulate = async (wrapper, selector, eventName) => {
  wrapper.find(selector).simulate(eventName);
  await forceUpdate(wrapper);
};

const currentLocation = (wrapper) => wrapper.find('MemoryRouter Router').prop('history').location;

/** Bring up the app with in-memory routing on enzyme/jsdom.
 * @param {string} route - the app route to load (see src/config/routes.js)
 * @param {object} options - optional settings
 * @param {boolean} options.authenticated - log into the app?
 * @param {boolean} options.clearApiAfterAuth - clear out the mock axios instance after auth?
 * @returns {Promise<WrappedApp>} an Enzyme wrapper containing the app
 */
export default async function mountRoute(route, { authenticated = true, clearApiAfterAuth = true } = {}) {
  const store = configureStore();

  // Note: axios#create() is mocked and now returns the same object on every call
  const axiosMock = axios.create();

  if (authenticated) {
    await store.dispatch(authenticate('test-username', 'test-password'));
    await store.dispatch(initializeAccessControl());

    if (clearApiAfterAuth) {
      axiosMock.mockClear();
    }
  }

  const Router = ({ children }) => <MemoryRouter initialEntries={[route]}>{children}</MemoryRouter>;

  const wrapper = mount(
    <Provider store={store}>
      <App RouterComponent={Router} />
    </Provider>
  );

  await asyncFlush();
  wrapper.update();

  /** A testable instance of the app
   * @typedef {object} WrappedApp
   * @property {ReactWrapper} wrapper - the underlying Enzyme wrapper
   * @property {function} currentRoute
   * @property {function} simulate - ReactWrapper#simulate
   * @property {function} forceUpdate - wait a tick and force the app to re-render
   * @property {object[]} mockApiCalls - array of calls made to axios
   * @property {object} axiosMock - the underlying (singleton) mock axios instance
   * @property {Store} store - the app's redux store
   */
  return {
    wrapper,
    currentLocation: () => currentLocation(wrapper),
    currentRoute: () => currentLocation(wrapper).pathname,
    find: wrapper.find.bind(wrapper),
    simulate: (...rest) => simulate(wrapper, ...rest),
    forceUpdate: (...rest) => forceUpdate(wrapper, ...rest),
    mockApiCalls: axiosMock.mock.calls,
    axiosMock,
    store
  };
}
