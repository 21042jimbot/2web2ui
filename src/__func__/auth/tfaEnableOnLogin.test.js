import mountRoute from 'src/__testHelpers__/mountRoute';
import getFormFiller from 'src/__testHelpers__/fill';

test('TFA flow: enable on login', async () => {
  const page = await mountRoute('/auth', { authenticated: false });

  const formFiller = getFormFiller(page.wrapper);

  formFiller([
    { name: 'username', value: 'tfa-required-username' },
    { name: 'password', value: 'tfa-required-password' }
  ]);

  await page.simulate('form', 'submit');
  expect(page.currentRoute()).toEqual('/auth/enable-tfa');

  formFiller({ selector: 'input', value: '314159' });

  await page.simulate('button[type="submit"]', 'click');
  expect(page.currentRoute()).toEqual('/dashboard');
  expect(page.mockApiCalls).toMatchSnapshot();
});
