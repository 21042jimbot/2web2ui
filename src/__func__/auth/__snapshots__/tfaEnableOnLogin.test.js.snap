// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TFA flow: enable on login 1`] = `
Array [
  Array [
    Object {
      "data": "grant_type=password&username=tfa-required-username&password=tfa-required-password&rememberMe=false",
      "headers": Object {
        "Authorization": "Basic bXN5c1dlYlVJOmZhODZkNzJlLTYyODctNDUxMy1hZTdmLWVjOGM4ZmEwZDc2Ng==",
        "Content-Type": "application/x-www-form-urlencoded",
      },
      "method": "POST",
      "url": "/v1/authenticate",
    },
  ],
  Array [
    Object {
      "data": "grant_type=password&username=tfa-required-username&password=tfa-required-password&rememberMe=false",
      "headers": Object {
        "Authorization": "Basic bXN5c1VJTGltaXRlZDphZjE0OTdkYS02NjI5LTQ3NTEtODljZS01ZDBmODE4N2MyMDQ=",
        "Content-Type": "application/x-www-form-urlencoded",
      },
      "method": "POST",
      "url": "/v1/authenticate",
    },
  ],
  Array [
    Object {
      "headers": Object {
        "Authorization": "mock-access-token",
      },
      "method": "GET",
      "url": "/v1/users/tfa-required-username/two-factor",
    },
  ],
  Array [
    Object {
      "data": Object {
        "enabled": true,
      },
      "headers": Object {
        "Authorization": "mock-access-token",
      },
      "method": "put",
      "params": undefined,
      "responseType": "json",
      "url": "/v1/users/tfa-required-username/two-factor",
    },
  ],
  Array [
    Object {
      "data": Object {
        "code": "314159",
        "enabled": true,
      },
      "headers": Object {
        "Authorization": "mock-access-token",
      },
      "method": "put",
      "params": undefined,
      "responseType": "json",
      "url": "/v1/users/tfa-required-username/two-factor",
    },
  ],
  Array [
    Object {
      "data": undefined,
      "headers": Object {
        "Authorization": "mock-access-token",
      },
      "method": "get",
      "params": Object {},
      "responseType": "json",
      "url": "/v1/account",
    },
  ],
  Array [
    Object {
      "data": undefined,
      "headers": Object {
        "Authorization": "mock-access-token",
      },
      "method": "get",
      "params": undefined,
      "responseType": "json",
      "url": "/v1/billing/plans",
    },
  ],
  Array [
    Object {
      "data": undefined,
      "headers": Object {
        "Authorization": "mock-access-token",
      },
      "method": "get",
      "params": Object {},
      "responseType": "json",
      "url": "/v1/billing/bundles",
    },
  ],
  Array [
    Object {
      "data": undefined,
      "headers": Object {
        "Authorization": "mock-access-token",
      },
      "method": "get",
      "params": Object {},
      "responseType": "json",
      "url": "/v1/billing/subscription",
    },
  ],
  Array [
    Object {
      "data": undefined,
      "headers": Object {
        "Authorization": "mock-access-token",
      },
      "method": "get",
      "params": undefined,
      "responseType": "json",
      "url": "/v1/users/tfa-required-username",
    },
  ],
  Array [
    Object {
      "data": undefined,
      "headers": Object {
        "Authorization": "mock-access-token",
      },
      "method": "get",
      "params": Object {
        "beta": false,
        "role": undefined,
      },
      "responseType": "json",
      "url": "/v1/authenticate/grants",
    },
  ],
  Array [
    Object {
      "data": undefined,
      "headers": Object {
        "Authorization": "mock-access-token",
      },
      "method": "get",
      "params": Object {
        "include": "usage",
      },
      "responseType": "json",
      "url": "/v1/account",
    },
  ],
  Array [
    Object {
      "data": undefined,
      "headers": Object {
        "Authorization": "mock-access-token",
        "x-msys-subaccount": 0,
      },
      "method": "get",
      "params": undefined,
      "responseType": "json",
      "url": "/v1/api-keys",
    },
  ],
  Array [
    Object {
      "data": undefined,
      "headers": Object {
        "Authorization": "mock-access-token",
      },
      "method": "get",
      "params": undefined,
      "responseType": "json",
      "url": "/v1/sending-domains",
    },
  ],
]
`;
